### Математические операции

In [1]:
# импорт и настройка
# импортирование библиотек
import pandas as pd
import numpy as np
pd.options.display.max_rows = 10
sp500 = pd.read_csv("../../source/sp500_2.csv", 
                    index_col='Symbol',
                    usecols=['Symbol','Sector','Price','BooK Value'])

In [2]:
sp500.head(2)

Sector   Price  BooK Value
Symbol                                 
MMM     Industrials  222.89       11.34
AOS     Industrials   60.24        6.35

In [3]:
# считываем исторические данные (датафрейм содержит данные об ациях компании mirosoft, apple)

In [4]:
omh = pd.read_csv("../../source/omh.csv",
                  parse_dates=['Date'])
omh.set_index('Date', inplace=True)

In [5]:
omh.head()

MSFT    AAPL
Date                     
2014-12-01  48.62  115.07
2014-12-02  48.46  114.63
2014-12-03  48.08  115.93
2014-12-04  48.84  115.49
2014-12-05  48.42  115.00

### Сводка статистик

In [58]:
# вычисляем сводку статистик для отдельного столбца Price (вычисление итоговой статистики для каждого столбца)

In [59]:
sp500.describe()

Price   BooK Value
count   505.000000   497.000000
mean    103.830634    14.453179
std     134.427636    89.660508
min       2.820000     0.510000
25%      46.250000     2.020000
50%      73.920000     3.400000
75%     116.540000     6.110000
max    1806.060000  1403.380000

|          |Price      | BooK Value | Description                                                           |
|----------|-----------|------------| ----------------------------------------------------------------------|  
| count    |505.000000 |497.000000  | значнея заполненые NoN - не считаются (поэтому разница в значениях )  |
| mean     |103.830634 |14.453179   | среднее значнение                                                     |                 
| std      |134.42763  |89.660508   | стандартное отклонение                                                |
| min      |2.820000   |0.510000    |                                                                       |
| 25%      |46.250000  |2.02000     | 25 - перцентиль                                                       |
| 50%      |73.920000  |3.400000    | 50 - перцентиль                                                       |
| 75%      |116.540000 |6.110000    | 75 - перцентиль                                                       | 
| max      |1806.060000|1403.380000 |                                                                       | 

In [7]:
sp500.Price.describe()

count     505.000000
mean      103.830634
std       134.427636
min         2.820000
25%        46.250000
50%        73.920000
75%       116.540000
max      1806.060000
Name: Price, dtype: float64

In [60]:
# получаем сводку статистик для нечисловых данных

|          |                      | Description                           |
|----------|----------------------|---------------------------------------|  
| count    |505                   |                                       |
| unique   |11                    |                                       |       
| top      |Consumer Discretionary| аиболее часто встречающиеся значение  |
| freq     |84                    |сколько раз встречается значение       |                                                                       |

In [67]:
sp500.Sector.describe()

count                        505
unique                        11
top       Consumer Discretionary
freq                          84
Name: Sector, dtype: object

In [10]:
# info - метод (кол-во и тип значений, объем памяти)

In [11]:
sp500.info()

<class 'pandas.core.frame.DataFrame'>
Index: 505 entries, MMM to ZTS
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Sector      505 non-null    object 
 1   Price       505 non-null    float64
 2   BooK Value  497 non-null    float64
dtypes: float64(2), object(1)
memory usage: 15.8+ KB


In [12]:
# получаем среднюю сводку для нечисловых данных

In [68]:
sp500.Sector.value_counts()

Consumer Discretionary        84
Information Technology        70
Financials                    68
Industrials                   67
Health Care                   61
                              ..
Real Estate                   33
Energy                        32
Utilities                     28
Materials                     25
Telecommunication Services     3
Name: Sector, Length: 11, dtype: int64

In [69]:
# получить долю 
sp500.Sector.value_counts(normalize=True)

Consumer Discretionary        0.166337
Information Technology        0.138614
Financials                    0.134653
Industrials                   0.132673
Health Care                   0.120792
                                ...   
Real Estate                   0.065347
Energy                        0.063366
Utilities                     0.055446
Materials                     0.049505
Telecommunication Services    0.005941
Name: Sector, Length: 11, dtype: float64

### Арифметические операции

- задаем стартовое значение генератора случайных чисел для получения воспроизводимых результатов
- созадем объект DataFrame

In [14]:
np.random.seed(123)
df = pd.DataFrame(np.random.randn(5, 4),
                  columns = ['A', 'B', 'C', 'D'])
df

A         B         C         D
0 -1.085631  0.997345  0.282978 -1.506295
1 -0.578600  1.651437 -2.426679 -0.428913
2  1.265936 -0.866740 -0.678886 -0.094709
3  1.491390 -0.638902 -0.443982 -0.434351
4  2.205930  2.186786  1.004054  0.386186

In [15]:
# умножаем все на 2, берем абсолютные значения
# получается новый датаферйм (исходный остается неизменным)

In [16]:
abs(df*2)

A         B         C         D
0  2.171261  1.994691  0.565957  3.012589
1  1.157201  3.302873  4.853358  0.857825
2  2.531873  1.733481  1.357772  0.189418
3  2.982779  1.277804  0.887964  0.868703
4  4.411860  4.373572  2.008108  0.772373

In [70]:
# При выполнении операции над DataFrame и Series 
# происходит выравниеваение Series по столбцам DataFrame

In [ ]:
# вычитание первой строки из каждой строки объекта DataFrame

In [18]:
df

A         B         C         D
0 -1.085631  0.997345  0.282978 -1.506295
1 -0.578600  1.651437 -2.426679 -0.428913
2  1.265936 -0.866740 -0.678886 -0.094709
3  1.491390 -0.638902 -0.443982 -0.434351
4  2.205930  2.186786  1.004054  0.386186

In [19]:
df.iloc[0]

A   -1.085631
B    0.997345
C    0.282978
D   -1.506295
Name: 0, dtype: float64

In [20]:
df - df.iloc[0]

A         B         C         D
0  0.000000  0.000000  0.000000  0.000000
1  0.507030  0.654091 -2.709658  1.077382
2  2.351567 -1.864086 -0.961865  1.411586
3  2.577020 -1.636247 -0.726960  1.071943
4  3.291561  1.189441  0.721075  1.892481

In [21]:
# вычитаем объект DataFrame из Series 

In [22]:
df.iloc[0] - df

A         B         C         D
0  0.000000  0.000000  0.000000  0.000000
1 -0.507030 -0.654091  2.709658 -1.077382
2 -2.351567  1.864086  0.961865 -1.411586
3 -2.577020  1.636247  0.726960 -1.071943
4 -3.291561 -1.189441 -0.721075 -1.892481

- возьмем второе и третье поле 1-ой строки:
- добавляем столбец Е
- смотрим, как применяется выравнивание в этой математической операции

In [23]:
df

A         B         C         D
0 -1.085631  0.997345  0.282978 -1.506295
1 -0.578600  1.651437 -2.426679 -0.428913
2  1.265936 -0.866740 -0.678886 -0.094709
3  1.491390 -0.638902 -0.443982 -0.434351
4  2.205930  2.186786  1.004054  0.386186

In [71]:
# создаем обект Series на основе 1 строки DataFrame В С + добавеляем Е
s = df.iloc[0][1:3]
s['E'] = 0
s

B    0.997345
C    0.282978
E    0.000000
Name: 0, dtype: float64

In [ ]:
# выравниеваение Series по столбцам DataFrame

In [25]:
df + s

A         B         C   D   E
0 NaN  1.994691  0.565957 NaN NaN
1 NaN  2.648782 -2.143701 NaN NaN
2 NaN  0.130605 -0.395908 NaN NaN
3 NaN  0.358443 -0.161003 NaN NaN
4 NaN  3.184132  1.287032 NaN NaN

In [26]:
# извлекаем строки в позициях с 1-й по 3-ю и только столбцы В и С
# по сути - извлекаем небольшой квадрат из середины df

In [27]:
subframe = df[1:4][['B', 'C']].copy()
subframe

B         C
1  1.651437 -2.426679
2 -0.866740 -0.678886
3 -0.638902 -0.443982

In [28]:
# как происходит выравнивание при выполнении операции вычитания

In [29]:
df - subframe

A    B    C   D
0 NaN  NaN  NaN NaN
1 NaN  0.0  0.0 NaN
2 NaN  0.0  0.0 NaN
3 NaN  0.0  0.0 NaN
4 NaN  NaN  NaN NaN

In [30]:
# извлекаем столбец А и вычитаем его из нашего датафрейма
df

A         B         C         D
0 -1.085631  0.997345  0.282978 -1.506295
1 -0.578600  1.651437 -2.426679 -0.428913
2  1.265936 -0.866740 -0.678886 -0.094709
3  1.491390 -0.638902 -0.443982 -0.434351
4  2.205930  2.186786  1.004054  0.386186

In [31]:
df.sub(df['A'], axis =0)

A         B         C         D
0  0.0  2.082976  1.368609 -0.420664
1  0.0  2.230037 -1.848079  0.149688
2  0.0 -2.132677 -1.944822 -1.360645
3  0.0 -2.130292 -1.935372 -1.925741
4  0.0 -0.019144 -1.201876 -1.819744

### Одномерные статистики

In [32]:
# минимум / максимум
# определяем максимальную цену для обеих акций

In [33]:
omh[['MSFT','AAPL']].min()

MSFT     45.16
AAPL    106.75
dtype: float64

In [34]:
# определим индекс, которому соответсвует максимальная цена для обеих акций

In [35]:
omh[['MSFT','AAPL']].idxmin()

MSFT   2014-12-16
AAPL   2014-12-16
dtype: datetime64[ns]

In [36]:
# среднее значение / медиана / мода

![среднее значение / медиана / мода](../../source/1.png)

In [37]:
# вычисление среднего значения для всех столбцов в датафрейме omh

In [38]:
omh.mean()

MSFT     47.493182
AAPL    112.411364
dtype: float64

In [39]:
# вычисление значения, усредненного по всем столбцам, для каждой строки (выводим только первые 5)

In [40]:
omh.mean(axis = 1).head()

Date
2014-12-01    81.845
2014-12-02    81.545
2014-12-03    82.005
2014-12-04    82.165
2014-12-05    81.710
dtype: float64

In [41]:
# вычисляем медиану значений для каждолго столбца

In [42]:
omh.median()

MSFT     47.625
AAPL    112.530
dtype: float64

In [43]:
# вычисляем моду для столбца Sector

In [44]:
sp500.Sector.mode()

0    Consumer Discretionary
dtype: object

In [45]:
# мод может быть несколько, пэтому результат операции - Series

In [46]:
s = pd.Series([1, 2, 3, 3, 5, 1])
s.mode()

0    1
1    3
dtype: int64

### дисперсия / среднеквадратичное отклонение

In [47]:
# вычисляем дисперсию значений в каждом столбце

In [48]:
omh.var()

MSFT    0.870632
AAPL    5.706231
dtype: float64

In [49]:
(omh.MSFT**2 - omh.MSFT.mean()**2).sum() / (omh.shape[0]-1)

0.8706322510824867

In [50]:
# вычисление среднеквадратичного отклонения

In [51]:
omh.std()

MSFT    0.933077
AAPL    2.388772
dtype: float64

In [52]:
omh.MSFT.var()**0.5

0.9330767659106356

###  ковариация / корреляция

In [53]:
# вычисляем ковариацию между MSFT и AAPL

In [54]:
omh.MSFT.cov(omh.AAPL)

1.9261240259740264

In [55]:
omh.MSFT.cov(omh.AAPL) / (omh.MSFT.std() * omh.AAPL.std())

0.8641560684381169

In [56]:
# либо можем получать матрицу ковариаций

In [57]:
omh.corr()

MSFT      AAPL
MSFT  1.000000  0.864156
AAPL  0.864156  1.000000